File: driver-service/src/main/java/com/dulfinne/taxi/driverservice/service/impl/DriverServiceImpl.java
package com.dulfinne.taxi.driverservice.service.impl;

import com.dulfinne.taxi.driverservice.exception.EntityNotFoundException;
import com.dulfinne.taxi.driverservice.exception.EntityAlreadyExistsException;
import com.dulfinne.taxi.driverservice.dto.request.DriverRequest;
import com.dulfinne.taxi.driverservice.dto.response.DriverResponse;
import com.dulfinne.taxi.driverservice.mapper.DriverMapper;
import com.dulfinne.taxi.driverservice.model.Car;
import com.dulfinne.taxi.driverservice.model.Driver;
import com.dulfinne.taxi.driverservice.repository.CarRepository;
import com.dulfinne.taxi.driverservice.repository.DriverRepository;
import com.dulfinne.taxi.driverservice.service.DriverService;
import com.dulfinne.taxi.driverservice.util.ExceptionKeys;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class DriverServiceImpl implements DriverService {

  private final DriverRepository driverRepository;
  private final CarRepository carRepository;
  private final DriverMapper driverMapper;

  @Transactional(readOnly = true)
  @Override
  public Page<DriverResponse> getAllDrivers(Integer offset, Integer limit, String sortField) {
    Page<Driver> driversPage =
        driverRepository.findAll(
            PageRequest.of(offset, limit, Sort.by(Sort.Direction.ASC, sortField)));

    return driversPage.map(driverMapper::toResponse);
  }

  @Transactional(readOnly = true)
  @Override
  public DriverResponse getDriverByUsername(String username) {
    Driver driver = getDriverIfExistByUsername(username);

    return driverMapper.toResponse(driver);
  }

  @Transactional
  @Override
  public DriverResponse saveDriver(String username, DriverRequest driverRequest) {
    checkUsernameUniqueness(username);
    checkPhoneNumberUniqueness(driverRequest.phoneNumber());

    Driver driver = driverMapper.toEntity(driverRequest);
    driver.setUsername(username);
    driver.setSumOfRatings(5.0);
    driver.setNumberOfRatings(1);
    driver = driverRepository.save(driver);
    return driverMapper.toResponse(driver);
  }

  @Transactional
  @Override
  public DriverResponse updateDriver(String username, DriverRequest driverRequest) {
    Driver driver = getDriverIfExistByUsername(username);
    checkPhoneNumberUniqueness(driver.getPhoneNumber(), driverRequest.phoneNumber());

    driverMapper.updateEntity(driverRequest, driver);
    driverRepository.save(driver);
    return driverMapper.toResponse(driver);
  }

  @Transactional
  @Override
  public void deleteDriver(String username) {
    Driver driver = getDriverIfExistByUsername(username);
    driverRepository.delete(driver);
  }

  @Transactional
  @Override
  public DriverResponse assignCarToDriver(String username, Long carId) {
    Driver driver = getDriverIfExistByUsername(username);
    Car car = getCarIfExist(carId);
    checkCarNotAssigned(carId);

    driver.setCar(car);
    driverRepository.save(driver);

    return driverMapper.toResponse(driver);
  }

  @Transactional
  @Override
  public DriverResponse removeCarFromDriver(String username) {
    Driver driver = getDriverIfExistByUsername(username);
    driver.setCar(null);
    driverRepository.save(driver);

    return driverMapper.toResponse(driver);
  }

  private Driver getDriverIfExistByUsername(String username) {
    return driverRepository
        .findByUsername(username)
        .orElseThrow(
            () -> new EntityNotFoundException(ExceptionKeys.DRIVER_NOT_FOUND_USERNAME, username));
  }

  private Car getCarIfExist(Long id) {
    return carRepository
        .findById(id)
        .orElseThrow(() -> new EntityNotFoundException(ExceptionKeys.CAR_NOT_FOUND_ID, id));
  }

  private void checkCarNotAssigned(Long carId) {
    if (driverRepository.findByCarId(carId).isPresent()) {
      throw new EntityAlreadyExistsException(ExceptionKeys.CAR_ALREADY_ASSIGNED, carId);
    }
  }

  private void checkUsernameUniqueness(String username) {
    if (driverRepository.findByUsername(username).isPresent()) {
      throw new EntityAlreadyExistsException(ExceptionKeys.DRIVER_EXISTS_USERNAME, username);
    }
  }

  private void checkPhoneNumberUniqueness(String phoneNumber) {
    if (driverRepository.findByPhoneNumber(phoneNumber).isPresent()) {
      throw new EntityAlreadyExistsException(ExceptionKeys.DRIVER_EXISTS_PHONE_NUMBER, phoneNumber);
    }
  }

  private void checkPhoneNumberUniqueness(String phoneNumber, String updatedPhoneNumber) {
    if (!updatedPhoneNumber.equals(phoneNumber)
        && driverRepository.findByPhoneNumber(updatedPhoneNumber).isPresent()) {
      throw new EntityAlreadyExistsException(
          ExceptionKeys.DRIVER_EXISTS_PHONE_NUMBER, updatedPhoneNumber);
    }
  }
}

File: ride-service/src/main/kotlin/com/dulfinne/taxi/rideservice/service/impl/DriverServiceImpl.kt
package com.dulfinne.taxi.rideservice.service.impl

import com.dulfinne.taxi.avro.Rating
import com.dulfinne.taxi.rideservice.client.service.ClientService
import com.dulfinne.taxi.rideservice.dto.request.RatingRequest
import com.dulfinne.taxi.rideservice.dto.response.AvailableRideResponse
import com.dulfinne.taxi.rideservice.dto.response.RideResponse
import com.dulfinne.taxi.rideservice.exception.ActionNotAllowedException
import com.dulfinne.taxi.rideservice.exception.EntityNotFoundException
import com.dulfinne.taxi.rideservice.kafka.service.KafkaProducerService
import com.dulfinne.taxi.rideservice.mapper.RideMapper
import com.dulfinne.taxi.rideservice.model.Ride
import com.dulfinne.taxi.rideservice.model.RideStatus
import com.dulfinne.taxi.rideservice.repository.RideRepository
import com.dulfinne.taxi.rideservice.service.DriverService
import com.dulfinne.taxi.rideservice.util.ExceptionKeys
import com.dulfinne.taxi.rideservice.util.RideConstants
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Duration
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.TimeZone

@Service
class DriverServiceImpl(
    val repository: RideRepository,
    val mapper: RideMapper,
    val kafkaService: KafkaProducerService,
    val clientService: ClientService
) : DriverService {

    @Transactional(readOnly = true)
    override fun getAvailableRides(username: String, radius: Int): List<AvailableRideResponse> {
        val point = mapper.toPoint(clientService.getDriverLocation(username))
        val list = repository.findByStatusAndRadius(RideStatus.SEARCHING.id, point, radius)
        return list.map(mapper::toAvailableRideResponse)
    }

    @Transactional
    override fun acceptRide(rideId: Long, driverUsername: String): RideResponse {
        val driver = clientService.getDriverByUsername(driverUsername)
        if (driver.car == null) {
            throw ActionNotAllowedException(ExceptionKeys.START_NOT_ALLOWED_CAR)
        }

        val ride = getRideIfExists(rideId)
        if (ride.status != RideStatus.SEARCHING.id) {
            throw ActionNotAllowedException(ExceptionKeys.ACCEPT_NOT_ALLOWED, RideStatus.fromId(ride.status))
        }

        ride.apply {
            this.driverUsername = driverUsername
            status = RideStatus.ACCEPTED.id
        }
        repository.save(ride)
        return mapper.toRideResponse(ride)
    }

    @Transactional
    override fun startRide(rideId: Long, driverUsername: String): RideResponse {
        val ride = getRideIfExists(rideId)
        validateDriver(ride, driverUsername)

        if (ride.status != RideStatus.ACCEPTED.id) {
            throw ActionNotAllowedException(ExceptionKeys.START_NOT_ALLOWED, RideStatus.fromId(ride.status))
        }

        ride.apply {
            startTime = ZonedDateTime.now(ZoneId.of(TimeZone.getDefault().id))
            status = RideStatus.IN_PROGRESS.id
        }
        repository.save(ride)
        return mapper.toRideResponse(ride)
    }

    @Transactional
    override fun finishRide(rideId: Long, driverUsername: String): RideResponse {
        val ride = getRideIfExists(rideId)
        validateDriver(ride, driverUsername)

        if (ride.status != RideStatus.IN_PROGRESS.id) {
            throw ActionNotAllowedException(ExceptionKeys.FINISH_NOT_ALLOWED, RideStatus.fromId(ride.status))
        }

        ride.apply {
            endTime = ZonedDateTime.now(ZoneId.of(TimeZone.getDefault().id))
            status = RideStatus.COMPLETED.id
        }
        repository.save(ride)
        return mapper.toRideResponse(ride)
    }

    @Transactional
    override fun ratePassenger(rideId: Long, driverUsername: String, request: RatingRequest) {
        val ride = getValidatedRideToRate(rideId, driverUsername)

        val rating = Rating.newBuilder().apply {
            username = ride.passengerUsername
            rating = request.rating
            feedback = request.feedback
        }.build()

        updateRideStatus(ride)
        repository.save(ride)

        kafkaService.sendPassengersRating(rating)
    }

    @Transactional(readOnly = true)
    override fun getAllDriverRides(
        driverUsername: String,
        offset: Int,
        limit: Int,
        sortField: String
    ): Page<RideResponse> {
        val ridesPage = repository.findAllByDriverUsername(
            driverUsername,
            PageRequest.of(offset, limit, Sort.by(Sort.Direction.ASC, sortField))
        )
        return ridesPage.map(mapper::toRideResponse)
    }

    @Transactional(readOnly = true)
    override fun getRideById(driverUsername: String, rideId: Long): RideResponse {
        val ride = getRideIfExists(rideId)
        validateDriver(ride, driverUsername)
        return mapper.toRideResponse(ride)
    }

    private fun checkRideCanBeRated(endTime: ZonedDateTime) {
        val duration = Duration.between(endTime, ZonedDateTime.now(ZoneId.of(TimeZone.getDefault().id))).toDays()
        if (duration > RideConstants.RATE_TIME_LIMIT_DAYS) {
            throw ActionNotAllowedException(ExceptionKeys.RATE_TIME_IS_OVER)
        }
    }

    private fun validateDriver(ride: Ride, driverUsername: String) {
        if (ride.driverUsername != driverUsername) {
            throw EntityNotFoundException(ExceptionKeys.RIDE_NOT_FOUND_ID, ride.id!!)
        }
    }

    private fun getRideIfExists(rideId: Long): Ride {
        return repository.findById(rideId).orElseThrow {
            EntityNotFoundException(ExceptionKeys.RIDE_NOT_FOUND_ID, rideId)
        }
    }

    private fun getValidatedRideToRate(rideId: Long, driverUsername: String): Ride {
        val ride = getRideIfExists(rideId)
        validateDriver(ride, driverUsername)
        if (ride.status != RideStatus.COMPLETED.id
            && ride.status != RideStatus.RATED_BY_PASSENGER.id
        ) {
            throw ActionNotAllowedException(ExceptionKeys.RATE_NOT_ALLOWED, RideStatus.fromId(ride.status))
        }

        checkRideCanBeRated(ride.endTime!!)
        return ride
    }

    private fun updateRideStatus(ride: Ride) {
        ride.status = when (ride.status) {
            RideStatus.COMPLETED.id -> RideStatus.RATED_BY_DRIVER.id
            RideStatus.RATED_BY_PASSENGER.id -> RideStatus.RATED.id
            else -> ride.status
        }
    }
}